src/com/hms_networks/americas/sc/time/LocalTimeOffsetCalculator.java

```diff
@@ -0,0 +1,138 @@
+      throws IOException, EWException, ParseException {
+    // Generate local time offset files
+    generateLocalTimeOffsetFiles();
+
+    // Parse local time offset value file and store/return
+    timeOffsetMilliseconds = parseOffsetFromLocalTime();
+    return timeOffsetMilliseconds;
+  }
+
+  /**
+   * Generates a local time offset HTML files and creates an HTTP connection to that page to
+   * generate a local time offset result file. The HTTP connection is authenticated using the
+   * default Ewon device credentials, or updated device credentials that can be set via {@link
+   * #setDeviceUsername(String)} and {@link #setDevicePassword(String)}.
+   *
+   * @throws IOException if unable to write local time offset HTML file
+   * @throws EWException if unable to create HTTP connection to local time offset HTML file (check
+   *     device username and password)
+   */
+  private static synchronized void generateLocalTimeOffsetFiles() throws IOException, EWException {
+    // Write local time offset HTML file
+    FileAccessManager.writeStringToFile(
+        TimeConstants.TIME_OFFSET_HTML_FILE_NAME, TimeConstants.TIME_OFFSET_HTML_FILE_CONTENTS);
+
+    // Load and write local time offset value file
+    String httpUserCredentialsAndServer = deviceUsername + ":" + devicePassword + "@127.0.0.1";
+    ScheduledActionManager.GetHttp(
+        httpUserCredentialsAndServer,
+        TimeConstants.TIME_OFFSET_RESULT_FILE_NAME,
+        TimeConstants.TIME_OFFSET_HTML_FILE_NAME);
+  }
+
+  /**
+   * Parses and returns the value from the local time offset result file generated by {@link
+   * #generateLocalTimeOffsetFiles()}.
+   *
+   * @return local time offset
+   * @throws IOException if unable to read local time offset result file
+   * @throws ParseException if unable to parse local time offset result file
+   */
+  private static long parseOffsetFromLocalTime() throws IOException, ParseException {
+    // Read contents of local time offset result file
+    BufferedReader reader =
+        new BufferedReader(new FileReader(TimeConstants.TIME_OFFSET_RESULT_FILE_NAME));
+    String line = reader.readLine();
+    reader.close();
+
+    // Remove unnecessary line break from file, if present
+    int endIndex = line.indexOf("<BR>");
+    if (endIndex > 0) {
+      line = line.substring(0, endIndex);
+    }
+
+    // Get local time from contents of local time offset result file
+    SimpleDateFormat sdf = new SimpleDateFormat(TimeConstants.TIME_OFFSET_DATE_FORMAT);
+    Date localTimeDateObj = sdf.parse(line);
+
+    // Calculate difference between UTC time (Ewon system time) and local time (in milliseconds)
+    Date systemTimeDateObj = new Date(System.currentTimeMillis());
+    long diffInMilliseconds = 0;
+    if (localTimeDateObj != null) {
+      diffInMilliseconds = systemTimeDateObj.getTime() - localTimeDateObj.getTime();
+    }
+
+    return diffInMilliseconds;
+  }
+}
```

Comment:
Since handling this error is probably going to be done is a fairly generic way should the library handle the case where the file cannot be accessed or parsed rather than making the application do it? This could just be another method that calls parseOffsetFromLocalTime() and handles the error.

Reply:
I'm not sure what you mean. In the event that the file cannot be accessed or parsed, the application would receive that exception (throws) and handle it by calling #calculateLocalTimeOffsetMilliseconds() again.

Reply:
Right but why have the application handle it when the way you handle it is fairly well defined. 

The code to call calculateLocalTimeOffsetMilliseconds() again (up to a certain number of attempts)  can exist in the library. Then the application really only needs to handle the case where there is a fatal error of not being able to access/create the file (file system corrupt or user credentials are wrong). this way this retry code is not replicated for every connector.

Reply:
The connectors currently don't retry. If the offset calculation fails, an error is shown and the offset is considered to be 0 milliseconds.